File: lib/message.ml
open! Core
module B = Bytes

module Request = struct
  type t =
    { client_id : int
    ; request_number : int
    ; op : Operation.t
    }
  [@@deriving bin_io, sexp, compare]
end

module Reply = struct
  type t =
    { view_number : int
    ; request_number : int
    ; result : Operation.result
    }
  [@@deriving bin_io, sexp, compare]
end

module Replica = struct
  type t =
    | Prepare of
        { view_number : int
        ; message : Request.t
        ; op_number : int
        ; commit_number : int
        }
    | PrepareOk of
        { view_number : int
        ; op_number : int
        ; replica_number : int
        }
  [@@deriving bin_io, sexp, compare]
end
(* | Reply of { view_number: int; } *)

type t =
  | Client_request of Request.t
  | Replica_message of Replica.t
[@@deriving bin_io, sexp]

let to_bytes req =
  let writer_msg = [%bin_writer: t] in
  let msg = Bin_prot.Utils.bin_dump ~header:true writer_msg req in
  let buf = B.create (Bin_prot.Common.buf_len msg) in
  Bin_prot.Common.blit_buf_bytes msg ~len:(B.length buf) buf;
  buf
;;

File: lib/message.rs
use std::fmt::Debug;

use quickcheck::{Arbitrary, Gen};
use serde::{Deserialize, Serialize};

use crate::{
  operation::{OpResult, Operation},
  types::{ClientID, CommitID, OpNumber, ReplicaID, RequestID, ViewNumber},
};

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
pub struct ClientRequest {
  pub client_id: ClientID,
  pub request_number: RequestID,
  pub op: Operation,
}

impl Arbitrary for ClientRequest {
  fn arbitrary(g: &mut Gen) -> Self {
    let client_id = Arbitrary::arbitrary(g);
    let request_number = Arbitrary::arbitrary(g);
    let op: Operation = Arbitrary::arbitrary(g);
    Self {
      client_id,
      request_number,
      op,
    }
  }
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
pub struct Reply {
  pub view_number: ViewNumber,
  pub request_number: RequestID,
  pub result: OpResult,
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
pub struct Prepare {
  pub view_number: ViewNumber,
  pub request: ClientRequest,
  pub op_number: OpNumber,
  pub commit_number: CommitID,
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
pub struct PrepareOk {
  pub view_number: ViewNumber,
  pub op_number: OpNumber,
  pub replica_number: ReplicaID,
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
pub enum ReplicaMessage {
  Prepare(Prepare),
  PrepareOk(PrepareOk),
}

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
pub enum IOMessage {
  Reply(Reply),
  Client(ClientRequest),
  Replica(ReplicaMessage),
}
